lib/stats/stats-query.c

```diff
@@ -218,77 +196,48 @@ _is_timestamp(gchar *counter_name)
 }
 
 void
-_sum_selected_counters(GList *counters, gpointer user_data)
+_sum_selected_counters(StatsCounterItem *counter, gpointer user_data, StatsFormatCb format_cb, gpointer result)
 {
-  GList *c;
   gpointer *args = (gpointer *) user_data;
   gint64 *sum = (gint64 *) args[1];
-  for (c = counters; c; c = c->next)
-    {
-      StatsCounterItem *counter = c->data;
-      if (!_is_timestamp(stats_counter_get_name(counter)))
-        *sum += stats_counter_get(counter);
-    }
+
+  if (!_is_timestamp(stats_counter_get_name(counter)))
+    *sum += stats_counter_get(counter);
 }
 
 gboolean
-_stats_query_get_sum(const gchar *expr, StatsSumFormatCb format_cb, gpointer result, gboolean must_reset)
+_stats_query_get_sum(const gchar *expr, StatsFormatCb format_cb, gpointer result, gboolean must_reset)
 {
   if (!expr)
     return FALSE;
 
-  gboolean found_match = FALSE;
+  const gchar *key_str = _setup_filter_expression(expr);
   gint64 sum = 0;
   gpointer args[] = {result, &sum};
+  gboolean found = _process_matching_counters(key_str, _sum_selected_counters, args, NULL, NULL, must_reset);
 
-  const gchar *key_str = _setup_filter_expression(expr);
-  GList *counters = _get_counters(key_str);
-  _sum_selected_counters(counters, (gpointer)args);
-
-  if (counters)
+  if (found)
     format_cb(args);
```

Comment:
I am not sure what format_cb() is here, but passing a gpointer array to a user-supplied function is far from being type safe or friendly to the caller. Why don't we just pass along the content of the args array as parameters?

Reply:
there are multiple kind of format callbacks here with different parameter sigantures, that is why i moved them into a single callback that using an args array
yes, it is not type safe, but there are multiple cases in the code that uses the same technic, a callback with an untyped args array